pkg/fqdn/dnsproxy/proxy.go

```diff
@@ -407,12 +407,12 @@ func (p *DNSProxy) SetRejectReply(opt string) {
 	}
 }
 
-// ExtractMsgDetails extracts a canonical query name, any IPs in a response and
-// the lowest applicable TTL. When a CNAME is returned the chain is collapsed
-// down, keeping the lowest TTL, and CNAME targets are returned.
-func ExtractMsgDetails(msg *dns.Msg) (qname string, responseIPs []net.IP, TTL uint32, CNAMEs []string, err error) {
+// ExtractMsgDetails extracts a canonical query name, any IPs in a response,
+// the lowest applicable TTL and message type. When a CNAME is returned the
+// chain is collapsed down, keeping the lowest TTL, and CNAME targets are returned.
+func ExtractMsgDetails(msg *dns.Msg) (qname string, responseIPs []net.IP, TTL uint32, CNAMEs []string, msgType int, recordTypes []uint16, err error) {
```

Comment:
I hadn't thought this far ahead but we also need the QType from the question (https://www.ietf.org/rfc/rfc1035.txt section 3.2.2 and 3.2.3, note that it is basically RRType with some extras). It will be relevant when no data is returned since that is the only indicator of what was asked. It's also possible to have a qtype like ANY that returns normal RRTypes.
It might be good to place it next to qname, since it goes with that data more clearly.